File: csharp-selenium-nunit-reqnroll/SauceDemo.UiTests/StepDefinitions/InventorySteps.cs
using OpenQA.Selenium;
using Reqnroll;
using SauceDemo.UiTests.Pages;
using SauceDemo.UiTests.TestData;

namespace SauceDemo.UiTests.StepDefinitions;

[Binding]
public sealed class InventorySteps
{
    private readonly IWebDriver _driver;
    private readonly ScenarioState _scenarioState;

    public InventorySteps(IWebDriver driver, ScenarioState scenarioState)
    {
        _driver = driver;
        _scenarioState = scenarioState;
    }

    [When(@"I add ""(.*)"" to the cart")]
    public void WhenIAddToTheCart(string productName)
    {
        new InventoryPage(_driver).AddToCart(productName);
        _scenarioState.ProductsInCart.Add(productName);
    }

    [When(@"I add the following products to the cart:")]
    public void WhenIAddTheFollowingProductsToTheCart(DataTable table)
    {
        var inventoryPage = new InventoryPage(_driver);

        foreach (var row in table.Rows)
        {
            var productName = row["ProductName"];
            inventoryPage.AddToCart(productName);
            _scenarioState.ProductsInCart.Add(productName);
        }
    }

    [When(@"I go to the cart")]
    public void WhenIGoToTheCart() => new InventoryPage(_driver).OpenCart();
}

File: python-selenium-pytest-bdd/src/saucedemo_uitests/steps/inventory_steps.py
from pytest_bdd import given, parsers, when
from selenium.webdriver.remote.webdriver import WebDriver

from saucedemo_uitests.pages.inventory_page import InventoryPage


@when(parsers.parse('I add "{product_name}" to the cart'))
def add_product_to_the_cart(driver: WebDriver, product_name: str) -> None:
    InventoryPage(driver).add_to_cart(product_name)


@when("I add the following products to the cart:")
def add_the_following_products_to_the_cart(driver: WebDriver, datatable: list[list[str]]) -> None:
    header, *rows = datatable
    name_index = header.index("ProductName")

    inventory_page = InventoryPage(driver)
    for row in rows:
        inventory_page.add_to_cart(row[name_index])


@when("I go to the cart")
def go_to_the_cart(driver: WebDriver) -> None:
    InventoryPage(driver).open_cart()


@given("my cart is empty")
def my_cart_is_empty(driver: WebDriver) -> None:
    InventoryPage(driver).clear_cart()
